### ISSUE TITLE
Improve resiliency of `auto-direct-node-routes` feature

### ISSUE THREAD

**Issue by gandro** [cilium/cilium#28137]
The `auto-direct-node-routes` feature of Cilium installs a route in the form of `$podCIDR via $nodeIP` for each remote node in the cluster, thereby allowing pod traffic to be routed to the connect node on a L2-connected network.

The current implementation lives in the `datapath.linuxNodeHandler` package:
https://github.com/cilium/cilium/blob/a4c43f358eee14c717f88b26386d458a6ddfd597/pkg/datapath/linux/node.go#L1337-L1345

It works by comparing the new and old version of the node object and updating the routes accordingly. This approach however has a major downside with more dynamic IPAM modes such as Multi-Pool: It misses updates to a node's podCIDRs that occur while the agent is down. Any route removed while the agent is down is likely leaked and not removed from the node until it is rebooted.

Therefore, a more resilient approach would be to follow a desired vs. realized state approach: The code should determine the desired set of routes and run a reconciliation loop to realize those routes, making sure all desired routes are actually installed on the system. The reconciliation loop would also be responsible to remove any routes not part of the desired set.
For this to work, we need to persistently store (i.e. persistent across cilium-agent restarts) what routes were previously installed. One potential solution could be to store the direct node routes in it's own routing table. That way, we assume that every route in that table is owned by the direct-node-routes controller.

---

*Update October 2023*: There is some ongoing work to implement a "routes reconciler" in https://github.com/cilium/cilium/pull/28276 that could form the basis for the work outlined in this issue.

Some observations from a discussion Jussi and David: The generic routes reconciler in that PR is a bit lower-level than what is needed here. In particular, the "desired routes" table in their work does not track routes provenance, something that we would want to have for auto-direct-node-routes (and likely also local-node-routes).

A way to approach this would be to have a separate "desired node routes" table that tracks node routes provenance, i.e. keeps a "owner reference" for each node route. This allows us then to remove all routes of a node if that node is removed or had it's pod CIDR changed. The flow of data would then be from a node route manager that tracks the owning node of each route, (living in or close to the existing `linuxNodeHandler`) to the "desired routes" table (which no longer has provenance), to the routes reconciler to system routing table.

**Comment by gandro** [comment]
@dylandreimerink Does #41528 fix this?

**Comment by dylandreimerink** [comment]
> This approach however has a major downside with more dynamic IPAM modes such as Multi-Pool: It misses updates to a node's podCIDRs that occur while the agent is down. Any route removed while the agent is down is likely leaked and not removed from the node until it is rebooted.

> Does https://github.com/cilium/cilium/pull/41528 fix this?

Yes. It persists an account of all routes that were installed via the route reconciler in a previous run. Upon startup, we wait to recreate the new desired state based on the new view of the world, then remove any old routes.

And the #41528 reconciler does track providence in the form of an owner, making it easy to remove all routes for a given node when it goes away for example.

We still need to switch `auto-direct-node-routes` over to using the route reconciler, since it currently still uses netlink directly. The reconciler is opt-in.

**Comment by dylandreimerink** [comment]
@andreaterzolo Turns out we did have an issue, here it is.

**Comment by andreaterzolo** [comment]
i would be happy to tackle this one :)
